### PR TITLE
[TACHYON-1092] Cleanup LocalTachyonCluster

### PR DESCRIPTION
**Pull request by apc999**
https://tachyon.atlassian.net/browse/TACHYON-1092

This PR aims to clean up code in `LocalTachyonCluster`, especially to provide a way for integration tests to modify the conf setup:

After this PR, integration tests use the following way to setup special test conf:

```java
localTachyonCluster =
    new LocalTachyonCluster(WORKER_CAPACITY_BYTES, QUOTA_UNIT_BYTES, BLOCK_SIZE_BYTES);
TachyonConf testConf = localTachyonCluster.getTestConf();
testConf.set(Constants.USER_FILE_BUFFER_BYTES, String.valueOf(BUFFER_BYTES));
testConf.set(Constants.WORKER_DATA_SERVER, MY_NETTY_DATA_SERVER);
// ... do more setting
localTachyonCluster.start()
```
When `localTachyonCluster` launching the worker and master, it uses the testConf as the basic to build configuration for both `WorkerContext`, `MasterContext`. This PR will prevent conf properties to propagate across different test cases.

### CODE REVIEW

File: clients/unshaded/src/main/java/tachyon/client/ClientContext.java [hunk 1]
```diff
@@ -143,5 +143,21 @@ public static void accessReinitializer(ReinitializerAccesser accesser) {
     accesser.receiveAccess(new PrivateReinitializer());
   }
 
-  private ClientContext() {}
+  /**
+   * Reset the TachyonConf instance in worker context, for test only.
+   * TODO(binfan): consider a better way to mock test TachyonConf
+   */
+  public static void resetConf() {
+    sTachyonConf = new TachyonConf();
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/ClientContext.java]: Do other static variables need to be reinitialized (master address?)

Comment (uronce-cc) [inline clients/unshaded/src/main/java/tachyon/client/ClientContext.java]: It seems that it is just to reset the TachyonConf without modifying other fields.

**Comment by uronce-cc** [comment]
+1 LGTM, I'm thinking about letting `LocalTachyonClusterMultiMaster` and `LocalTachyonCluster` share a `LocalTachyonClusterBase`.

File: integration-tests/src/test/java/tachyon/client/IsolatedTachyonFileSystemIntegrationTest.java [hunk 2]
```diff
@@ -56,23 +56,23 @@ public final void after() throws Exception {
...
+
+    mLocalTachyonCluster.init();
+    MasterContext.getConf().set(Constants.USER_FILE_BUFFER_BYTES, Integer.toString(
+        USER_QUOTA_UNIT_BYTES));
     mLocalTachyonCluster.start();
 
     mTfs = mLocalTachyonCluster.getClient();
 
-    TachyonConf mWorkerTachyonConf;
+    TachyonConf workerTachyonConf;
```
Comment (yupeng9) [inline integration-tests/src/test/java/tachyon/client/IsolatedTachyonFileSystemIntegrationTest.java]: NIT: This line and the one below can be combined.

File: servers/src/main/java/tachyon/master/MasterContext.java [hunk 3]
```diff
@@ -55,10 +55,20 @@ public static MasterSource getMasterSource() {
...
   /**
-   * Reset the TachyonConf instance in master context, for test only.
+   * Resets the master context, for test only.
    * TODO(binfan): consider a better way to mock test TachyonConf
    */
-  public static void resetConf() {
-    sTachyonConf = new TachyonConf();
+  public static void reset() {
+    reset(new TachyonConf());
+
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/MasterContext.java]: extra blank line?

File: servers/src/main/java/tachyon/worker/WorkerContext.java [hunk 4]
```diff
@@ -40,10 +40,18 @@ public static TachyonConf getConf() {
   }
 
   /**
-   * Reset the TachyonConf instance in worker context, for test only.
+   * Resets the worker context, for test only.
    * TODO(binfan): consider a better way to mock test TachyonConf
    */
-  public static void resetConf() {
+  public static void reset() {
     sTachyonConf = new TachyonConf();
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/WorkerContext.java]: should this just call the other reset() method, like in MasterContext?

**Comment by gpang** [comment]
This still seems a bit confusing, since the test has to get the correct context conf to set the right values. What if there was a static method getTestConf() that returns a test conf (with everything in init()) and tests can add to that conf any other variables it wants. Then tests can call start(conf) with that modified conf (or the unmodified test conf).

This would be a way for the test writer to not care about which conf it needs to update. Just some thoughts...

**Comment by uronce-cc** [comment]
+1 for gene's suggestion, so we just call `getTestConf` and set the properties, then call `start`, the internal test conf will be set into master/worker/client conf. It's more straightforward.

**Comment by jsimsa** [comment]
@apc999  what is the status of this PR? should it be closed?

File: servers/src/main/java/tachyon/worker/WorkerContext.java (hunk 4 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/WorkerContext.java]: done

File: servers/src/main/java/tachyon/master/MasterContext.java (hunk 3 above)
Comment (apc999) [inline servers/src/main/java/tachyon/master/MasterContext.java]: removed.

File: integration-tests/src/test/java/tachyon/client/IsolatedTachyonFileSystemIntegrationTest.java (hunk 2 above)
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/IsolatedTachyonFileSystemIntegrationTest.java]: done

**Comment by apc999** [comment]
@gpang , I followed your suggestion to provide testConf. @jsimsa , @gpang @uronce-cc , please take a look. There will be followup PRs to further clean `LocalTachyonMaster` by shifting some work to `LocalTachyonCluster`.

**Comment by apc999** [comment]
@uronce-cc Having integration test base is good. We can work on that after this PR.

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java [hunk 5]
```diff
@@ -74,39 +72,38 @@ public final void after() throws Exception {
...
+  public static void beforeClass() throws IOException {
+    sLocalTachyonCluster =
+        new LocalTachyonCluster(WORKER_CAPACITY_BYTES, QUOTA_UNIT_BYTES, BLOCK_SIZE_BYTES);
+    sTestConf = sLocalTachyonCluster.getTestConf();
+    sTestConf.set(Constants.USER_FILE_BUFFER_BYTES, String.valueOf(BUFFER_BYTES));
+    // Only the Netty data server supports remote writes.
+    sTestConf.set(Constants.WORKER_DATA_SERVER, IntegrationTestConstants.NETTY_DATA_SERVER);
     sWriteBoth =
-        new OutStreamOptions.Builder(mMasterTachyonConf)
+        new OutStreamOptions.Builder(sTestConf)
```
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: Hmm I think passing in conf to the builder is a mistake if we are assuming the conf is the same as `ClientContext.getConf`, but this is definitely not in the scope of this PR.

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java [hunk 6]
```diff
@@ -166,8 +163,8 @@ public void writeTest1() throws IOException, TachyonException {
     }
   }
 
-  private void writeTest1Util(TachyonURI filePath, int len, OutStreamOptions op) throws IOException,
-      TachyonException {
+  private void writeTest1Util(TachyonURI filePath, int len, OutStreamOptions op)
+      throws IOException, TachyonException {
```
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: Is this the new style? (`throws` on next line)

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 5 above)
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: sgtm

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 6 above)
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: this is done by formatter. Does tachyon have a new style file recently?

Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: oh okay, my formatter must not be updated.

File: integration-tests/src/test/java/tachyon/client/IsolatedTachyonFileSystemIntegrationTest.java [hunk 7]
```diff
@@ -56,23 +55,21 @@ public final void after() throws Exception {
...
+    testConf.set(Constants.USER_FILE_BUFFER_BYTES, Integer.toString(
+        USER_QUOTA_UNIT_BYTES));
     mLocalTachyonCluster.start();
 
     mTfs = mLocalTachyonCluster.getClient();
 
-    TachyonConf mWorkerTachyonConf;
-
-    mWorkerTachyonConf = mLocalTachyonCluster.getWorkerTachyonConf();
-    mWorkerTachyonConf.set(Constants.MAX_COLUMNS, "257");
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/IsolatedTachyonFileSystemIntegrationTest.java]: no longer needed?

File: integration-tests/src/test/java/tachyon/client/TachyonFileSystemIntegrationTest.java [hunk 8]
```diff
@@ -55,17 +54,12 @@
...
-  private TachyonConf mMasterTachyonConf;
-  private TachyonConf mWorkerTachyonConf;
 
   @Rule
   public ExpectedException mThrown = ExpectedException.none();
 
   @Before
   public final void before() throws IOException, TException {
-    mMasterTachyonConf = sLocalTachyonCluster.getMasterTachyonConf();
-    mMasterTachyonConf.set(Constants.MAX_COLUMNS, "257");
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/TachyonFileSystemIntegrationTest.java]: no longer needed?

File: servers/src/main/java/tachyon/master/MasterContext.java [hunk 9]
```diff
@@ -55,10 +55,19 @@ public static MasterSource getMasterSource() {
   }
 
   /**
-   * Reset the TachyonConf instance in master context, for test only.
+   * Resets the master context, for test only.
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/MasterContext.java]: Can we avoid adding a public methods that are used in tests only? I think we should be able to use PowerMock for this. (See BlockDataManagerTest for an example of extensive use of mocking).

File: servers/src/main/java/tachyon/worker/WorkerContext.java [hunk 10]
```diff
@@ -40,10 +40,18 @@ public static TachyonConf getConf() {
   }
 
   /**
-   * Reset the TachyonConf instance in worker context, for test only.
+   * Resets the worker context, for test only.
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/worker/WorkerContext.java]: ditto

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 11]
```diff
@@ -53,22 +69,25 @@ public static void main(String[] args) throws Exception {
...
   private BlockWorker mWorker = null;
   private long mWorkerCapacityBytes;
   private int mUserBlockSize;
   private int mQuotaUnitBytes;
   private String mTachyonHome;
-  private String mWorkerDataFolder;
   private Thread mWorkerThread = null;
   private String mLocalhostName = null;
   private LocalTachyonMaster mMaster;
   private TachyonConf mMasterConf;
```
Comment (jsimsa) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: do we still need the individual confs (Master, Worker, Client)?

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 12]
```diff
@@ -53,22 +69,25 @@ public static void main(String[] args) throws Exception {
...
   private TachyonConf mWorkerConf;
+  private TachyonConf mClientConf;
+  private TachyonConf mTestConf;
```
Comment (jsimsa) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: Could we just call it mTachyonConf?

**Comment by jsimsa** [comment]
@apc999 thank you for this PR, it is a great improvement! 

Left some comments for you. Also, please resolve conflicts.

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 13]
```diff
@@ -115,88 +134,117 @@ public NetAddress getWorkerAddress() {
...
     // Perform immediate shutdown of data server. Graceful shutdown is unnecessary and slow
-    mWorkerConf.set(Constants.WORKER_NETWORK_NETTY_SHUTDOWN_QUIET_PERIOD, Integer.toString(0));
-    mWorkerConf.set(Constants.WORKER_NETWORK_NETTY_SHUTDOWN_TIMEOUT, Integer.toString(0));
+    mTestConf.set(Constants.WORKER_NETWORK_NETTY_SHUTDOWN_QUIET_PERIOD, Integer.toString(0));
+    mTestConf.set(Constants.WORKER_NETWORK_NETTY_SHUTDOWN_TIMEOUT, Integer.toString(0));
 
     // Since tests are always running on a single host keep the resolution timeout low as otherwise
     // people running with strange network configurations will see very slow tests
-    mWorkerConf.set(Constants.NETWORK_HOST_RESOLUTION_TIMEOUT_MS, Integer.toString(250));
+    mTestConf.set(Constants.NETWORK_HOST_RESOLUTION_TIMEOUT_MS, Integer.toString(250));
```
Comment (uronce-cc) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: duplicates with https://github.com/amplab/tachyon/pull/1713/files#diff-8163899588eea6fa1c089e0cf00ed3d2R157

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 14]
```diff
@@ -115,88 +134,117 @@ public NetAddress getWorkerAddress() {
...
+   *
+   * @throws IOException when the operation fails
+   */
+  private void startWorker() throws IOException {
+    mWorkerConf = new TachyonConf(mTestConf.getInternalProperties());
+    mClientConf = new TachyonConf(mTestConf.getInternalProperties());
+    WorkerContext.reset(mWorkerConf);
+    ClientContext.reset(mClientConf);
 
+    // We need to update the client with the most recent configuration so it knows the correct ports
```
Comment (uronce-cc) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: This comment seems to be for the above codes?

Comment (uronce-cc) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: `correct ports` of what? If the `ClientContext` only needs to update its knowledge of `MASTER_PORT`, why not put related logic into `startMaster()`?

File: integration-tests/src/test/java/tachyon/client/TachyonFileSystemIntegrationTest.java (hunk 8 above)
Comment (uronce-cc) [inline integration-tests/src/test/java/tachyon/client/TachyonFileSystemIntegrationTest.java]: Yes, this is for raw table, it's unrelated to this test.

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 6 above)
Comment (uronce-cc) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: I often use this style, it looks pretty to me.

**Comment by uronce-cc** [comment]
@apc999 It makes the configuration much much cleaner now! To clean up `LocalTachyonClusterMultiMaster`, maybe we can let it extend a base class: https://tachyon.atlassian.net/browse/TACHYON-1199

File: integration-tests/src/test/java/tachyon/client/IsolatedTachyonFileSystemIntegrationTest.java (hunk 7 above)
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/IsolatedTachyonFileSystemIntegrationTest.java]: I believe so

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 11 above)
Comment (apc999) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: It's possible to have all conf instance pointing to the same object. I still keep them separately for two reasons: (1) technically, those conf should be in different JVM and independent from each other; (2) we may want to add tests in the future where the conf diverge. Storing them separately enables these kinds of test.

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 13 above)
Comment (apc999) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: Thanks. removed.

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 12 above)
Comment (apc999) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: I thought `mTestConf` is more informative than `mTachyonConf`? It is kinda obvious this `conf` is a Tachyon Configuration, and I thought `mTestConf` makes the name aligned with `m{Master,Worker,Client}Conf`

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 14 above)
Comment (apc999) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: done

File: servers/src/main/java/tachyon/master/MasterContext.java (hunk 9 above)
Comment (apc999) [inline servers/src/main/java/tachyon/master/MasterContext.java]: Agree with avoiding using public methods only for testing. But I wish to address this in a separate PR (and try a different approach from whitebox). My concern with simply using whitebox in this particular case is that the re-initialize logic will be moved to the integration test which might be miss match in the future if we add new stuff into context class but forget to update the integration tests.

File: servers/src/main/java/tachyon/worker/WorkerContext.java (hunk 10 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/WorkerContext.java]: same concern as stated above.

**Comment by apc999** [comment]
@uronce-cc yes, SGTM.

**Comment by apc999** [comment]
PTAL, thanks!

**Comment by apc999** [comment]
Jenkins, test this please.

File: integration-tests/src/test/java/tachyon/client/TachyonFileSystemUtilsIntegrationTest.java [hunk 15]
```diff
@@ -73,14 +59,14 @@ public static final void afterClass() throws Exception {
...
   @BeforeClass
   public static void beforeClass() throws Exception {
-    MasterContext.getConf().set(Constants.USER_FILE_BUFFER_BYTES, Integer.toString(
-        USER_QUOTA_UNIT_BYTES));
     sLocalTachyonCluster =
       new LocalTachyonCluster(WORKER_CAPACITY_BYTES, USER_QUOTA_UNIT_BYTES, Constants.GB);
+    sLocalTachyonCluster.getTestConf().set(Constants.USER_FILE_BUFFER_BYTES, Integer.toString(
+        USER_QUOTA_UNIT_BYTES));
+    sLocalTachyonCluster.getTestConf().set(Constants.MAX_COLUMNS, "257");
+
```
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/TachyonFileSystemUtilsIntegrationTest.java]: Is this necessary?

File: integration-tests/src/test/java/tachyon/client/table/RawTableIntegrationTest.java [hunk 16]
```diff
@@ -49,8 +48,8 @@ public final void after() throws Exception {
 
   @Before
   public final void before() throws Exception {
-    MasterContext.getConf().set(Constants.USER_FILE_BUFFER_BYTES, String.valueOf(100));
     mLocalTachyonCluster = new LocalTachyonCluster(10000, 1000, Constants.GB);
+    mLocalTachyonCluster.getTestConf().set(Constants.USER_FILE_BUFFER_BYTES, String.valueOf(100));
```
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/table/RawTableIntegrationTest.java]: Should the hard coded numbers be constants?

File: integration-tests/src/test/java/tachyon/master/JournalIntegrationTest.java [hunk 17]
```diff
@@ -157,7 +157,7 @@ public final void after() throws Exception {
   @Before
   public final void before() throws Exception {
     mLocalTachyonCluster = new LocalTachyonCluster(Constants.GB, 100, Constants.GB);
-    MasterContext.getConf().set(Constants.MASTER_JOURNAL_LOG_SIZE_BYTES_MAX,
+    mLocalTachyonCluster.getTestConf().set(Constants.MASTER_JOURNAL_LOG_SIZE_BYTES_MAX,
```
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/master/JournalIntegrationTest.java]: I prefer this style when only 1 entry needs to be set, could you make it consistent for the other cases?

File: integration-tests/src/test/java/tachyon/client/TachyonFileSystemUtilsIntegrationTest.java (hunk 15 above)
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/TachyonFileSystemUtilsIntegrationTest.java]: removed.

File: integration-tests/src/test/java/tachyon/client/table/RawTableIntegrationTest.java (hunk 16 above)
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/table/RawTableIntegrationTest.java]: This is from original code, and I think we should not freeride this PR for too much unrelated minor fix. We can probably create small tasks on this later.

File: integration-tests/src/test/java/tachyon/master/JournalIntegrationTest.java (hunk 17 above)
Comment (apc999) [inline integration-tests/src/test/java/tachyon/master/JournalIntegrationTest.java]: Done.

File: integration-tests/src/test/java/tachyon/client/table/RawTableIntegrationTest.java (hunk 16 above)
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/table/RawTableIntegrationTest.java]: SGTM

**Comment by calvinjia** [comment]
LGTM
@apc999 I think this is a good step to cleaning up the tech debt, but there's still several things that should be addressed in subsequent PRs.

**Comment by apc999** [comment]
@calvinjia , yes https://tachyon.atlassian.net/browse/TACHYON-1092 is actually an umbrella JIRA and you can find or create more related JIRA under TACHYON-1092

**Pull request merged**